### PR TITLE
Command line argument parsing improvements

### PR DESCRIPTION
**Pull request by FrankRay78**
The following three PR's, when taken jointly, represent a significant enhancement to existing command line parsing behaviour: https://github.com/spectreconsole/spectre.console/pull/1036, https://github.com/spectreconsole/spectre.console/pull/1029, https://github.com/spectreconsole/spectre.console/pull/1040

However, they all make changes to CommandTreeTokenizer and introduce their own unit tests specifically for the CommandTreeTokenizer, and as such, merge conflicts will arise once any one of the PR's has been merged.

This PR encompasses all three PR's, successfully merged and with all conflicts resolved. The respective git histories have also been squashed so the changes for each issue can be more clearly examined. 

All issues addressed by this single combined PR are: https://github.com/spectreconsole/spectre.console/issues/959, https://github.com/spectreconsole/spectre.console/issues/842, https://github.com/spectreconsole/spectre.console/issues/890, https://github.com/spectreconsole/spectre.console/issues/193 (and its open duplicate, https://github.com/spectreconsole/spectre.console/issues/825), https://github.com/spectreconsole/spectre.console/issues/587

As it's quite a big change, I'm happy to help in any way required @patriksvensson for you to accept in full, in part, with or without amendment etc.

### CODE REVIEW

**Comment by patriksvensson** [comment]
@FrankRay78 From a quick look, it looks ok to me! I will set aside some time this weekend to dig into it a bit more.

Does this one cover other pull requests? If so, could you close the other ones to avoid confusion?

Thanks!

File: src/Spectre.Console.Cli/Internal/Parsing/CommandTreeTokenizer.cs [hunk 1]
```diff
@@ -29,7 +29,14 @@ public static CommandTreeTokenizerResult Tokenize(IEnumerable<string> args)
         var context = new CommandTreeTokenizerContext();
 
         foreach (var arg in args)
-        {
+        {
+            if (arg == string.Empty)
```
Comment (patriksvensson) [inline src/Spectre.Console.Cli/Internal/Parsing/CommandTreeTokenizer.cs:33]: Can we change this to `string.IsNullOrEmpty(arg)`?

**Review by patriksvensson — CHANGES_REQUESTED** [review]
A minor thing, but otherwise, it looks good to me 👍 

@phil-scott-78, @nils-a Anything you would want to be addressed before merging?

**Comment by nils-a** [comment]
> Anything you would want to be addressed before merging?

Nothing that you haven't addressed already. 👍

File: src/Spectre.Console.Cli/Internal/Parsing/CommandTreeTokenizer.cs (hunk 1 above)
Comment (FrankRay78) [reply src/Spectre.Console.Cli/Internal/Parsing/CommandTreeTokenizer.cs:33]: Thanks, I will address this over the weekend @patriksvensson.

**Review by patriksvensson — APPROVED** [review]

**Comment by patriksvensson** [comment]
@FrankRay78 There are currently conflicts with the main branch. Could you rebase your PR branch with an updated main branch?

If you're uncertain how to do this, we've got a guide for it in our [Contribution Guidelines](https://github.com/spectreconsole/spectre.console/blob/main/CONTRIBUTING.md#respond-to-feedback-on-pull-request)

**Comment by FrankRay78** [comment]
I've rebased and fixed the merge conflicts @patriksvensson.  It all seems good, however I get the following message (which I'm not convinced is unrelated to the merge):

> Build started...
1>------ Build started: Project: Spectre.Console, Configuration: Debug Any CPU ------
1>C:\Users\info\.nuget\packages\tunnelvisionlabs.referenceassemblyannotator\1.0.0-alpha.160\build\TunnelVisionLabs.ReferenceAssemblyAnnotator.targets(58,5): error RA0000: Add a 'PackageDownload' project item to include Microsoft.NETCore.App.Ref at the desired version to obtain nullability information from.

I started seeing that on some of my other branches today after I refreshed them from upstream.

Any ideas how I fix this?

**Comment by patriksvensson** [comment]
@FrankRay78 Merged!

Thank you for this. Stellar work! Really happy with your work! 👍

**Pull request merged**